x-pack/plugins/elastic_assistant/server/conversations_data_client/append_conversation_messages.ts

```diff
@@ -0,0 +1,105 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ElasticsearchClient, Logger } from '@kbn/core/server';
+
+import {
+  ConversationResponse,
+  Message,
+  UUID,
+} from '../schemas/conversations/common_attributes.gen';
+import { getConversation } from './get_conversation';
+
+export const appendConversationMessages = async (
+  esClient: ElasticsearchClient,
+  logger: Logger,
+  conversationIndex: string,
+  userId: UUID,
+  existingConversation: ConversationResponse,
+  messages: Message[]
+): Promise<ConversationResponse | null> => {
+  const updatedAt = new Date().toISOString();
+
+  const params = transformToUpdateScheme(updatedAt, [
+    ...(existingConversation.messages ?? []),
+    ...messages,
+  ]);
+  try {
+    const response = await esClient.updateByQuery({
+      conflicts: 'proceed',
+      index: conversationIndex,
+      query: {
+        ids: {
+          values: [existingConversation.id ?? ''],
+        },
+      },
+      refresh: true,
+      script: {
+        lang: 'painless',
+        params: {
+          ...params,
+        },
+        source: `
+          if (params.assignEmpty == true || params.containsKey('messages')) {
+            def messages = [];
+            for (message in params.messages) {
+              def newMessage = [:];
+              newMessage['@timestamp'] = message['@timestamp'];
+              newMessage.content = message.content;
+              newMessage.is_error = message.is_error;
+              newMessage.presentation = message.presentation;
+              newMessage.reader = message.reader;
+              newMessage.replacements = message.replacements;
+              newMessage.role = message.role; 
+              messages.add(newMessage);
+            }
+            ctx._source.messages = messages;
+          }
+          ctx._source.updated_at = params.updated_at;
+        `,
+      },
+    });
+    if (response.failures && response.failures.length > 0) {
+      logger.warn(
+        `Error appending conversation messages: ${response.failures.map(
+          (f) => f.id
+        )} for conversation by ID: ${existingConversation.id}`
+      );
+      return null;
+    }
+  } catch (err) {
```

Comment:
Consider replacing the literal `Error` with `Failure` for consistency with `logger.warn`, if this path uses `logger.warn` because (for example), the operation may be retried.